indexed_gzip/zran_file_util.c

```diff
@@ -106,40 +106,28 @@ int64_t _ftell_python(PyObject *f) {
 int _fseek_python(PyObject *f, int64_t offset, int whence) {
 
     PyObject *data = NULL;
-    PyObject *seek_fn_name = NULL;
-    PyObject *whence_ = NULL;
-    PyObject *offset_ = NULL;
 
     _ZRAN_FILE_UTIL_ACQUIRE_GIL
 
-    // We can't use PyObject_CallMethod with multiple arguments
-    // because it causes tests with 32-bit OS wheels to fail
-    // (as the second argument is not passed correctly to Python),
-    // so we have to manually build up the arguments instead.
-    // TODO: File an issue with cpython about this.
-    // if ((data = PyObject_CallMethod(f, "seek", "(l,i)", offset, whence)) == NULL)
-    //     goto fail;
-    if ((seek_fn_name = PyUnicode_FromString("seek")) == NULL)
+    /*
+     * The seek method returns type long, which has
+     * different sizes on different platforms
+     */
+    if (sizeof(long) == 8)
+        data = PyObject_CallMethod(f, "seek", "(l,i)", offset, whence);
+    else if (sizeof(long long) == 8)
+        data = PyObject_CallMethod(f, "seek", "(L,i)", offset, whence);
```

Comment:
That's interesting. I saw your earlier comment about this @pauldmccarthy , but I'm still a bit unclear / curious as to how exactly this would fail: why would using `l` instead of `L` (giving the wrong type for the first argument) have any effect on the second argument?

Reply:
It's to do with how `PyObject_CallMethod` is un/re-marshalling the arguments - the bytes from `offset` are overflowing into the re-constructed value for `whence`. We're giving it an offset of type `int64_t`, but telling it that it is of type `l`, which corresponds to `long`, which is 4 bytes on a typical 32 bit platform. So the upper four bytes of `int64_t offset` are used to construct the value that is used as the offset passed to the python method, and the lower four bytes of `int64_t offset` are used as the whence passed to the python. The actual value of the `int whence` argument is not used at all.

It's a little bit crude, but you can see what is happening by making these changes to the latest implementation of `_fseek_python`:

1. Change `L` back to `l` on line 119
2. Clobber the value of `offset` before it is passed to `PyObject_CallMethod`, e.g.:
```c
// on a little endian architecture, this will cause 
// the top four bytes to contain (int32_t)54321, 
// and the bottom four bytes to contain (int32_t)12345
offset = ((int64_t)12345<<32) + (int32_t)54321;
```

Compile, and try to create an `IndexedGzipFile` from an in-memory file-like. e.g.:
```python
import io
import indexed_gzip as igzip
fobj = io.BytesIO(open('any_file.gz', 'rb').read())
# this will induce a call to fobj.seek
f = igzip.IndexedGzipFile(fileobj=fobj) 
```
I get:
```ValueError: invalid whence (12345, should be 0, 1 or 2)```

Reply:
That makes sense, thanks for the explanation!